File: apps/agent/internal/docker/ports.go
package docker

import (
	"strconv"

	"github.com/docker/docker/api/types/container"
)

// PortsFromList converts the port entries Docker returns for a listed
// container into the protocol shape: lowerCamelCase keys, the host port as a
// string that is empty when the port is exposed but not published, and the
// bound host interface when Docker reports one. The result is never nil so the
// wire always carries an array, and Docker's order is preserved.
func PortsFromList(ports []container.Port) []Port {
	result := make([]Port, 0, len(ports))
	for _, port := range ports {
		public := ""
		if port.PublicPort != 0 {
			public = strconv.Itoa(int(port.PublicPort))
		}
		result = append(result, Port{
			Private:  int(port.PrivatePort),
			Public:   public,
			Protocol: port.Type,
			IP:       port.IP,
		})
	}
	return result
}

File: apps/web/src/lib/ports.ts
import type { Port } from '@/types/api'

/**
 * Every port mapping of a container, joined rather than showing only the
 * first one. A port that is exposed but not published has an empty `public`
 * and renders as the private port alone, e.g. `8080:80, 443`.
 */
export function formatPorts(ports: Port[] | undefined): string {
  if (!ports || ports.length === 0) return '-'
  return ports
    .map((port) => (port.public ? `${port.public}:${port.private}` : `${port.private}`))
    .join(', ')
}
